### DISCUSSION TITLE
Lerna 7 Roadmap

### DISCUSSION THREAD

**Discussion by JamesHenry** [lerna/lerna#3410] [General]
We are delighted with what we've achieved with the help of the lerna community in the 6 months since we took over stewardship of Lerna!

For information about what lerna looks like today, including all the exciting stuff that was shipped in Lerna v6, please check out the following resources:

- https://blog.nrwl.io/lerna-reborn-whats-new-in-v6-10aec6e9091c
- https://blog.nrwl.io/upgrade-your-lerna-workspace-make-it-fast-and-modern-a8d17c4f95a2
- **Upgrade your Lerna Workspace - Make it Fast and Modern:** https://www.youtube.com/watch?v=MwGPMvhJS0A
- **Modern Lerna Walkthrough:** https://www.youtube.com/watch?v=1oxFYphTS4Y
- **Lerna Reborn talk from Nx Conf:** https://www.youtube.com/watch?v=CNdDv2MsBuw

Now let's turn our attention to the near future! 🔮 

**We will plan to do major releases of lerna about twice per year**, landing any number of minor and patch releases in between.

There you can expect Lerna 7 to be the latest version of Lerna around May 2023.

Please read on to see what we have planned for Lerna 7, and note that anything which does not require breaking changes in order to land will be available at some point in Lerna v6 minor releases.

---

# Lerna 7 Focus Areas

## `--dry-run` for `lerna version` and `lerna publish`

As the longest standing open feature request dating back over 6 years (https://github.com/lerna/lerna/issues/51), we are delighted to be able turn our attention to adding powerful `--dry-run` capabilities to both the `version` and `publish` commands.

This means that when you pass the `--dry-run` flag you will be able to see a preview of what those commands will do before those updates take effect on disk/over the network.

## Recoverable `lerna publish`

Something we have run into ourselves whilst publishing lots of packages is that if your internet connection is unstable, or if your target registry goes down, whilst running the `publish` command, you can end up in a state where some of your packages have been published, and some haven't.

Lerna currently does not give you an easy way to resolve this situation, and it is left to you to iterate through your packages and run the appropriate `npm publish` commands to tidy things up again.

In Lerna 7, we want to make it so that `lerna publish` can handle this scenario gracefully, and you won't need to manually run `npm publish` in this scenario any more!

## Workspace level watch mode for scripts

There is currently no tool that we know about which offers project graph aware file watching when running scripts across an entire workspace. There are plenty of tools which help you spin up lots of watch processes, but these do not ensure that the requisite dependencies between projects are correctly considered when choosing what to rebuild.

In Lerna 7, we are working on a solution for this, so that you can e.g. run the `build` script for all your packages in your workspace in `watch` mode, and have them rebuild in the appropriate and most efficient order, with the full power of the script caching (available by default since Lerna 6) in order to skip unnecessary work.

## Transitioning away from legacy package management features

Lerna is the original monorepo/workspace tool in the JavaScript ecosystem. When it was created in 2015/2016 the ecosystem looked totally different, and there were no built in capabilities to handle working with multiple packages in a single repository (a "workspace"). Commands like `lerna bootstrap`, `lerna add` and `lerna link` were all a critical part of the lerna project, because there were no other options.

However, now that we find ourselves in late 2022, the fact is that - for many years now - the package managers we know and love (`npm`, `yarn` and `pnpm`) all fully support that concept of workspaces as a first-class use-case.

They have battle tested implementations covering adding, removing and linking local packages, and combining them with third party dependencies in a natural way.

This is the reason why, for the past several _years_ of his tenure as lead maintainer of Lerna @evocateur has been encouraging folks to strongly reconsider their use of the legacy package management commands in lerna, and instead leverage their package manager of choice to do what it does best.

We knew about this context from afar, but as new stewards of the project we did not want to jump straight in and start removing capabilities without first taking the time to get familiar with the reality up close. Now that we have been actively maintaining the project for half a year, we are in full agreement with Daniel and others that the legacy package management commands in lerna need to be retired.

By removing these legacy pieces which have _better_ alternatives natively in package managers, we and the rest of the lerna community will be freed up to concentrate our efforts on things which are uniquely valuable about lerna (such as, but not limited to, versioning and publishing), and making them the best they can be!

From version 7, therefore, we will no longer ship the legacy package management commands (such as `lerna bootstrap`, `lerna add` and `lerna link`) by default with lerna, and instead make them available as an add-on via a separate package (name TBD).

This new package can be thought of as being in maintenance mode only - no new features will be considered for legacy package management concerns (such as `lerna bootstrap`, `lerna add` and `lerna link`), and we will only look to merge critical patches and security updates.

We of course want to make sure that folks who are perhaps less aware of the modern capabilities of their package manager are not left confused by this change, so we will be building out comprehensive migration guides on https://lerna.js.org to help them transition their thinking from a legacy lerna command to its equivalent for `npm`, `yarn` or `pnpm`.

We will also begin immediately communicating this on old issues and PRs related to these commands.

**Comment by gziolo** [comment]
> Lerna currently does not give you an easy way to resolve this situation, and it is left to you to iterate through your packages and run the appropriate npm publish commands to tidy things up again.

That’s going to be a great improvement to have it automatically handled. It isn’t that bad at the moment once you learn that you can  call `npx lerna publish from-package` (with the same additional arguments as the unfinished publishing) to resume publishing. However, it's often tricky to put the branch again in the proper state that lets Lerna do its magic.

**Comment by dep** [comment]
One critical component of `lerna bootstrap` that our team has leaned on is the ability to `nohoist` specific dependencies.  Would appreciate any guidance or documentation around how one might migrate from `lerna bootstrap` + selective `nohoist` to something like npm workspaces when you do start collecting resources or help for folks. Thanks!

**Reply by LitvinaviciusM** [reply]
Not sure if it's still relevant, but here's what worked for me 👉  https://classic.yarnpkg.com/blog/2018/02/15/nohoist/

**Comment by Bec-k** [comment]
Would be cool to have a function, which will stop managing all dependencies and move them back to each managed package.
This is very useful when you want to build Dockerfile inside some specific package, but it is not possible, because lerna is managing all dependencies at higher level (i.e. at the top node_modules folder)

Additional functionality which would be cool to have, is to pack all listed packages (not external, but the ones managed by lerna) in package.json and move them into node_modules, instead of putting fs links there.
This will be useful for Dockerfile as well.

Something like `lerna artifact @company/backend` or `lerna standalone @company/backend` and it will prepare package with all required for docker image building.

That's the problem we have been experiencing while working with lerna all this time. Implementing above workaround.

**Reply by JamesHenry** [reply]
Hi @denissabramovs the point of this change in lerna v7 is that `lerna` will no longer be attempting manage packages for you. That job is best left to package managers which all have support for workspaces.

If none of the package managers already implement the kind of functionality you are requesting then it may be worth opening a discussion with them to see if they are open to adding it.

If for whatever reason that ends up not being something which any of the package managers want to built support for, I would suggest building a prototype of it which meets your needs in isolation and then sharing that with folks here to see what work is actually needed on top of the package manager. Perhaps at that point it could be something which be considered as an extension to lerna, but that would need to be considered in light of the work required and how far it deviates from lerna's core mission.

Hopefully that makes sense!

**Reply by Bec-k** [reply]
Yes, possibly this could be done via yarn or npm or any other, which supports workspaces.

**Reply by jaequery** [reply]
@JamesHenry i am in the same boat as @dennisabramovs , how can we have NX to manage all package dependencies at per package level instead of all in one place?

**Comment by bwhitty** [comment]
Related to @dep's note about "no hoist" https://github.com/lerna/lerna/discussions/3410#discussioncomment-4079104, another issue with the removal of `lerna bootstrap` is the loss of per-package lockfiles. The issue is described in detail here https://github.com/npm/rfcs/issues/554, but I think it also has implications on `nx affected`:

If updating a single package's dependencies always yields a diff to the root `package-lock.json`, then won't `nx affected` always re-run again the entire monorepo since `package-lock.json` is an `input` to the `nx` task runner by default?

As [described in `yarn nohoist`'s introductory blog post](https://classic.yarnpkg.com/blog/2018/02/15/nohoist/), Lerna's support for this is explicitly called out:
> What is “nohoist” ?
>
> Is there a simple yet universal mechanism that can allow these incompatible libraries working in the monorepo environment?
>
> Turns out there is, and is conveniently called “nohoist”, which has also been demonstrated in other monorepo tools like [lerna](https://github.com/lerna/lerna/blob/master/doc/hoist.md).
>
> “nohoist” enables workspaces to consume 3rd-party libraries not yet compatible with its hoisting scheme. The idea is to disable the selected modules from being hoisted to the project root. They were placed in the actual (child) project instead, just like in a standalone, non-workspaces, project.
>
> Since most 3rd-party libraries already worked in standalone projects, the ability to simulate such environment within workspaces should be able to unblock many known compatibility issues.

When applied to something like `nx affected`, would the "fix" here to ignore the root `package-lock.json` and only use all the `package.json`s as "input"s to nx?
